### PR TITLE
New device project S905

### PR DESCRIPTION
**Pull request by adamg-xda**
This is a WIP as discussed with @Raybuntu 

I would invite the following reviewers to keep an eye on this and review as required
@afl1 @CGarces @kszaq

### CODE REVIEW

**Comment by jernejsk** [comment]
This driver will come handy for Allwinner boards, since some of them have this chip. What is your kernel target? For Allwinner branch, we use latest mainline.

**Comment by Raybuntu** [comment]
Thank you @adamg-xda.

**Comment by CGarces** [comment]
Hi @jernejsk 
The AmLogic builds use 3.14 and 3.10 so the driver will not work for mainline. If is needed, I can try to apply some patches, but note that I don't have that hardware.

**Comment by adamg-xda** [comment]
@CGarces I have the RTL8189FS so I can test it, this is why this is the only additional driver I have provided for now.

**Comment by Raybuntu** [comment]
@adamg-xda I'd like to ask you to please open seperate PR's as stated in https://github.com/LibreELEC/LibreELEC.tv/blob/master/CONTRIBUTING.md
You can reopen your driver PR and remove the commit and also the rng-tools.
It is easier to review and some changes can be merged before.

**Comment by jernejsk** [comment]
@CGarces 
no worries, I can help with maintaining patches for mainline. I was just hoping that we have same target.

**Comment by adamg-xda** [comment]
@Raybuntu will do that

**Comment by chewitt** [comment]
I'm currently 50/50 on whether this should be allowed as a project or should be restructured along similar lines to #2125, but I guess the advantage of project is we keep the 3.14 hackfest separate from our future desired (mainline) state, and this makes junking all Amlogic non-mainline projects easier at some future point. Sadly it appears this won't happen for Leia/LE9.0.

The "custom" kernel hack is acceptable in a private fork but is not acceptable in a release branch because it negates collaboration and version control which is one of the key objectives of adding this to our main repo. We can accept an alternate git repo URL until nougat sources are merged to our linux-amlogic repo - @codesnake is working on this. All linux patches should then be applied to the kernel repo for the benefit of all projects unless there is true justification for a project-specific patch. As this is a generic project I wouldn't expect anything to be required.

There should be no 'OEM' packages in a release, we are not an OEM.

Kodi patches are not acceptable unless items have been PR'd to Kodi and are waiting on upstream merge. In that scenario patch names should reflect the upstream PR number. Same for libcec.

Binary add-ons will be provided from our repo (under our version control) so please revert to our standard URL. Also remove the errant blank line in advancedsettings.xml and don't line \ break ADDITIONAL_DRIVERS in project options.

I'm sure there's more .. that's just a first pass :)

**Comment by MilhouseVH** [comment]
Agreed, the custom kernel change should be dropped as it's no longer necessary - just add the new kernel and any patches to the `linux` package.

I'm not sure about adding and then building the device tree in the `linux ` build directory - I'm wondering what this does for the kernel stamp, and may continually make it "dirty" meaning it is rebuilt with every incremental build unless the device tree cleans up after itself. I'm also wondering if `packages/devel/aml-s9xx-device-trees` is the right location, maybe `packages/linux-drivers/aml-s9xx-device-trees` would make more sense? It does however need:
```
PKG_NEED_UNPACK="$LINUX_DEPENDS"
```
so that the device tree is rebuilt if/when the kernel package is changed in any way.

**Comment by adamg-xda** [comment]
@codesnake I was very 50/50 on whether to submit this for this very scrutiny, the edit to kernel package is not a hack and version control is still maintained within projects/S905/options, it just saves filling up the file with multiple kernels like the old imx6 entries. In a perfect world the kernels would be unified but we are not quite there yet and I haven't even attempted to look into this yet. If you have a better suggestion then it would be more helpful if you say it rather than saying this is unacceptable because it's counter productive.

I was assured PRs here for S905 would be welcomed.

I will make the other changes mentioned and look at sending some of the patches to the respective upstreams.

**Comment by lrusak** [comment]
@adamg-xda we always welcome PR's from all contributors however you need to expect a level of scrutiny before it can be merged. Don't look at it as condescending but as positive criticism that will improve the project as a whole. 

The whole custom kernel thing will not be allowed. Sorry. This makes it very difficult to track sources  because one could set this option outside of the project options and in the custom options `~/.libreelec/options`. We need to be able to have a clear line of sight for what packages are being used for a build. This is important for making progress to reproducible builds.

So please just work with us here and we can get through the needed changes.

**Comment by chewitt** [comment]
@adamg-xda Someone had to submit the PR so don't take the comments personally :) ..and some scrutiny is always needed to realign things with project standards. If you only work in one area of the buildsystem keeping kernel versions 'separate' makes sense, but for those of us who need to keep track of all projects/devices etc. fragmenting things makes it a little harder and we work to avoid the death of a thousand cuts.

**Comment by adamg-xda** [comment]
@lrusak your comment makes sense, I've made some changes as requested.

I'm fine with constructive criticism and I welcome changes to bring it closer to LE standards.

**Comment by MilhouseVH** [comment]
The `aml-dtbtools` package is missing. Also, I'm not sure that adding the dependency to `toolchain` is the right place for this - what package needs whatever `aml-dtbtools` provides?

**Comment by adamg-xda** [comment]
There is more changes to come yet @MilhouseVH, it should be ready tomorrow night hopefully, I need to review and rewrite some other changes.

**Comment by kszaq** [comment]
@adamg-xda Thank you for your work on submitting this project to main LE repo! After a brief look I can tell that some changes to scripts are missing. If you don't mind, I will submit them as separate PRs shortly. I prefer to submit these changes one by one so that we don't have to review everything at once. :)

**Comment by adamg-xda** [comment]
@kszaq yeah I know I was tired last night and didn't have chance to finish it all, I have left most of the drivers out for additional wifi devices to be added at a later date as well as support for devices with dvb tuners. The idea was to get basic S905 into master that could be expanded upon.

**Comment by adamg-xda** [comment]
This now compiles and creates an image that boots on S905 however it has broken the buildsystem for other devices, The offending commit is #819e8a7 with the code in scripts/image, I will look into this later.

**Comment by Kwiboo** [comment]
Is there more AML projects incoming? Else `kodi-aml-hdmimonitor.sh/service` could be move to ` projects/S905/filesystem` like the rest of the project specific filesystem files

**Comment by adamg-xda** [comment]
@Kwiboo there is the S805 and S912 but I don't think anybody wants those given S805 is ageing and S912 relies on libhybris to work and other black magic/voodoo patches, so I will move it on my next push :)

**Comment by adamg-xda** [comment]
This now compiles and creates an image that boots on S905 however it has broken the buildsystem for other devices, The offending commit is a9757ed with the multi-dtb code in scripts/image.

I will wait for #2253 #2260 #2261 before proceeding further.

File: packages/mediacenter/kodi/package.mk [hunk 1]
```diff
@@ -345,4 +345,8 @@ post_install() {
   enable_service kodi-waitonnetwork.service
   enable_service kodi.service
   enable_service kodi-lirc-suspend.service
+
+  if [ $PROJECT = "S905" ]; then
+    enable_service kodi-aml-hdmimonitor.service
```
Comment (Kwiboo) [inline packages/mediacenter/kodi/package.mk]: You do not need this if you also move `kodi-aml-hdmimonitor.service` to `$PROJECT/filesystem/usr/lib/systemd/system`, all `.service`-files in such project/device path is enabled by the LE build system

File: projects/S905/filesystem/usr/share/bootloader/update.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
+#!/bin/sh
+
+################################################################################
+#      This file is part of LibreELEC - https://libreelec.tv
+#      Copyright (C) 2016 Team LibreELEC
```
Comment (CvH) [inline projects/S905/filesystem/usr/share/bootloader/update.sh]: Copyright (C) 2016-present Team LibreELEC

File: packages/virtual/toolchain/package.mk [hunk 3]
```diff
@@ -22,7 +22,7 @@ PKG_ARCH="any"
 PKG_LICENSE="GPL"
 PKG_SITE="https://libreelec.tv"
 PKG_URL=""
-PKG_DEPENDS_TARGET="configtools:host make:host xz:host sed:host pkg-config:host autoconf:host automake:host intltool:host libtool:host autoconf-archive:host gcc:host bison:host flex:host cmake:host xmlstarlet:host yasm:host p7zip:host ninja:host meson:host"
+PKG_DEPENDS_TARGET="configtools:host make:host xz:host sed:host pkg-config:host autoconf:host automake:host intltool:host libtool:host autoconf-archive:host gcc:host bison:host flex:host cmake:host xmlstarlet:host yasm:host p7zip:host ninja:host meson:host aml-dtbtools:host"
```
Comment (MilhouseVH) [inline packages/virtual/toolchain/package.mk]: This is still wrong - `aml-dtbtools` is not a toolchain dependency, it's a dependency of a specific aml package (`linux`?), so should be added there.

File: packages/mediacenter/kodi/package.mk (hunk 1 above)
Comment (adamg-xda) [reply packages/mediacenter/kodi/package.mk]: I didn't know this, thanks for the tip

**Comment by chewitt** [comment]
Why closed?

**Comment by Raybuntu** [comment]
Afaik it will be refactored and reopened.

**Pull request closed without merging**